tensorflow/core/kernels/dynamic_stitch_op.cc

```diff
@@ -20,9 +20,119 @@ limitations under the License.
+      auto indices_vec = indices.flat<int32>();
+      const Tensor& data = data_inputs[input_num];
+      auto data_flat =
+          data.shaped<T, 2>({indices_vec.dimension(0), slice_size});
+
+      if (DataTypeCanUseMemcpy(DataTypeToEnum<T>::v())) {
+        T* merged_base = &merged_flat(0, 0);
+        const T* data_base = &data_flat(0, 0);
+        const size_t slice_bytes = slice_size * sizeof(T);
+        for (int i = 0; i < indices_vec.size(); i++) {
+          int32 index = internal::SubtleMustCopy(indices_vec(i));
+          OP_REQUIRES(
+              c, FastBoundsCheck(index, first_dim_size),
+              errors::InvalidArgument("indices[", i, "] is out of range"));
+          memcpy(merged_base + index * slice_size, data_base + i * slice_size,
+                 slice_bytes);
+        }
+      } else {
+        Eigen::DSizes<Eigen::DenseIndex, 2> sizes(1, slice_size);
+        for (int i = 0; i < indices_vec.size(); i++) {
+          // Copy slice data[i] to merged[indices[i]]
+          Eigen::DSizes<Eigen::DenseIndex, 2> data_indices(i, 0);
+          int32 index = internal::SubtleMustCopy(indices_vec(i));
+          OP_REQUIRES(
+              c, FastBoundsCheck(index, first_dim_size),
+              errors::InvalidArgument("indices[", i, "] is out of range"));
+          Eigen::DSizes<Eigen::DenseIndex, 2> merged_indices(index, 0);
+          merged_flat.slice(merged_indices, sizes) =
+              data_flat.slice(data_indices, sizes);
+        }
+      }
+    }
+  }
+};
+
+#if GOOGLE_CUDA
+
+template <typename T>
+struct DynamicStitchGPULaunch {
+  static void Run(const GPUDevice& d, const int slice_size,
+                  const int first_dim_size,
+                  const CudaDeviceArrayStruct<int32>& indices_flat,
+                  const CudaDeviceArrayStruct<const T*>& data_slice_ptrs,
+                  T* output);
+};
+
+template <typename T>
+struct LaunchDynamicStitchOp<GPUDevice, T> {
+  static void launch(OpKernelContext* c, const int slice_size,
+                     const int first_dim_size, OpInputList indices_inputs,
+                     OpInputList data_inputs, Tensor* merged) {
+    // createt two arrays that will be sent to CUDA device
+    // one used as pointers to output's row id
+    CudaDeviceArrayOnHost<int32> indices_flat(c, first_dim_size);
+    // another used as pointers to the pointers of the head of each data_slice
+    CudaDeviceArrayOnHost<const T*> data_slice_ptrs(c, first_dim_size);
+    OP_REQUIRES_OK(c, indices_flat.Init());
+    OP_REQUIRES_OK(c, data_slice_ptrs.Init());
+
+    int ptr_num = 0;
+    for (int input_num = 0; input_num < indices_inputs.size(); input_num++) {
+      auto indices_vec = indices_inputs[input_num].flat<int32>();
+      auto base_ptr = data_inputs[input_num].template flat<T>().data();
+      for (int ind_num = 0; ind_num < indices_vec.size(); ind_num++) {
+        indices_flat.Set(ptr_num, indices_vec(ind_num));
```

Comment:
Yes, I should actually just replace `ptr_num` with the value of `indices_vec(ind_num)`, so when collision happens the new index value will naturally overwrite the old one, which would be the same as how current implementation handles collisions, is my understanding correct? 

However, if we end up taking this approach, we would also need to account for missing indices on `indices_flat`. To do this, I plan on adding 1 to all the values of output index then subtract it back on the GPU kernel, so if the threads on CUDA kernel encounter a 0 on `output_id` they can just skip the read and write. 

Unfortunately, I think `CudaDeviceArrayOnHost` is not zero initialized, so I would need to do it manually. Is there a standard way to zero-initialize a temporary tensor? or do I just use normal for-loop to achieve this?